File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

string solution(vector<int> numbers, string hand) {
    map<int,pair<int,int>> location;
    for(int i=1; i<13; i++){
        location.insert({i, make_pair((i-1)/3, (i-1)%3)});
    }
    
    string answer = "";
    
    int left = 10;
    int right = 12;
    
    for(int i=0; i<numbers.size(); i++){
        int cur = numbers[i];
        
        if(cur==0) cur=11;
        
        if(location[cur].second == 0){
            answer += "L";
            left = cur;
        }
        else if (location[cur].second == 2){
            answer += "R";
            right = cur;
        }
        else{
            int ldist = abs(location[left].first - location[cur].first) + abs(location[left].second - location[cur].second);
            int rdist = abs(location[right].first - location[cur].first) + abs(location[right].second - location[cur].second);
            if (ldist > rdist){
                answer += "R";
                right = cur;
            }else if (ldist == rdist){
                if (hand == "left"){
                    answer += "L";
                    left = cur;
                }else{
                    answer += "R";
                    right = cur;
                }
            }else{
                answer += "L";
                left = cur;
            }
        }
    }
    return answer;
}

File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    
    location = dict()
    for i in range(1,13):
        location[i] = [(i-1)//3, (i-1)%3]
    
    left = 10
    right = 12
    
    for i in numbers:
        if i == 0:
            i = 11
        if i in [1,4,7]:
            answer += "L"
            left = i
        elif i in [3,6,9]:
            answer += "R"
            right = i
        else:
            # 거리 계산
            ldist = abs(location[left][0] - location[i][0]) + abs(location[left][1] - location[i][1])
            rdist = abs(location[right][0] - location[i][0]) + abs(location[right][1] - location[i][1]) 
            
            # 키패드 누를 손가락 결정
            if ldist > rdist:
                answer += "R"
                right = i
            elif ldist == rdist:
                if hand == "right":
                    answer += "R"
                    right = i
                else:
                    answer += "L"
                    left = i
            else:
                answer += "L"
                left = i

    return answer